neural_compressor/compression/pruner/wanda/prune.py

```diff
@@ -0,0 +1,218 @@
+# !/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2022 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import transformers
+
+from neural_compressor.adaptor.torch_utils.util import get_hidden_states
+
+from .utils import find_layers, get_module_list, get_tensor_sparsity_ratio, logger, nn, torch
+from .wrapper import WrappedGPT
+
+
+@torch.no_grad()
+def prepare_calibration_input(model, dataloader, device):
+    use_cache = model.config.use_cache
+    model.config.use_cache = False
+    layers = get_module_list(model)
+
+    if hasattr(model, "hf_device_map") and "model.embed_tokens" in model.hf_device_map:
+        device = model.hf_device_map["model.embed_tokens"]
+
+    inps = []
+    others = []
+
+    class Catcher(nn.Module):
+        def __init__(self, module):
+            super().__init__()
+            self.module = module
+
+        def forward(self, *inp, **kwargs):
+            tmp_other = {}
+            for arg in kwargs:
+                if isinstance(kwargs[arg], torch.Tensor) or arg == "alibi":
+                    tmp_other[arg] = kwargs[arg]
+            others.append(tmp_other)
+            inps.append(list(inp))
+            raise ValueError
+
+    layers[0] = Catcher(layers[0])
+    for batch in dataloader:
+        try:
+            model(batch[0])
+        except ValueError:
+            pass
+    layers[0] = layers[0].module
+
+    model.config.use_cache = use_cache
+
+    return inps, others
+
+
+def return_given_alpha(alpha, sort_res, W_metric, tmp_metric, sum_before):
+    thres_cumsum = sum_before * alpha
+    sort_mask = tmp_metric <= thres_cumsum.reshape((-1, 1))
+    thres = torch.gather(sort_res[0], dim=1, index=sort_mask.sum(dim=1, keepdims=True) - 1)
+    W_mask = W_metric <= thres
+    cur_sparsity = torch.sum(W_mask is True).data.item() / W_mask.numel()
+    return W_mask, cur_sparsity
+
+
+def move_inps_to_device(inps, others, device):
+    for idx in range(len(inps)):
+        for i in range(len(inps[idx])):
+            inps[idx][i] = inps[idx][i].to(device)
+        for arg in others[idx]:
+            other = others[idx][arg]
+            if isinstance(other, torch.Tensor):
+                others[idx][arg] = other.to(device)
+
+
+@torch.no_grad()
+def prune_wanda(
+    model,
+    dataloader,
+    sparsity_ratio,
+    prune_n=0,
+    prune_m=0,
+    nsamples=128,
+    use_variant=False,
+    device=None,
+    low_mem_usage=False,
```

Comment:
Detecting the need for low memory usage in this case is not difficult. Could we set the default value to None and have it detected automatically?"

Reply:
Done